### PR TITLE
Fix `PushRuleEvaluator` and `Filter` to work on frozendicts

### PR DESCRIPTION
**Pull request by richvdh**
frozendicts do not (necessarily) inherit from dict, so this needs to handle them correctly.

It seems like the Python version of `frozendict` *does* inherit from `dict`, which is why we have gotten away with this for so long. Versions of `frozendict` since 2.2 make it easier to install the C version (there are now prebuilt `manylinux` wheels).

### CODE REVIEW

**Comment by clokep** [comment]
Seems to be a regression (by me) from #11098. 😢

**Comment by richvdh** [comment]
> Seems to be a regression (by me) from #11098.

Interesting, I hadn't spotted that that had changed.

Anyway yes, that looks like part of it, but you still need a few other things to go wrong before this bug actually bites. Hopefully the unit test will help here in future.

File: changelog.d/12100.bugfix [hunk 1]
```diff
@@ -0,0 +1 @@
+Fix a long-standing bug which could cause push notifications to malfunction if `use_frozen_dicts` was set in the configuration.
```
Comment (squahtx) [inline changelog.d/12100.bugfix:1]:
```suggestion
Fix a bug introduced in 1.47.0 which could cause push notifications to malfunction if `use_frozen_dicts` was set in the configuration.
```

**Review by squahtx — APPROVED** [review]
lgtm

**Comment by clokep** [comment]
(And the second one seems to be a regression from #11194?)

**Comment by richvdh** [comment]
> (And the second one seems to be a regression from https://github.com/matrix-org/synapse/pull/11194?)

sorry yes, I said that in the commit comment.

**Pull request merged**